akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala

```diff
@@ -557,6 +566,39 @@ private[akka] class ShardRegion(
   val initRegistrationDelay: FiniteDuration = 100.millis.max(retryInterval / 2 / 2 / 2)
   var nextRegistrationDelay: FiniteDuration = initRegistrationDelay
 
+  // FIXME not too happy about giving the shard actor context to the factory but the persistent one needs to spawn a child
```

Comment:
This is the thing I was talking about @patriknw 

Unhappy with (but can't come up with a better strategy for) accessing the shard ActorContext from a lambda elsewhere (needed to spawn persistent per shard as child), but needs to be defined here to allow for a single shared instance for the ddata one.

Also, the magic reflection one only needed for testing is not nice. Sharding starting through an extension and not spawning the individual actors manually since they expect certain interaction and that would make for really brittle tests (there were a few I changed away from doing that)

Reply:
does it have to be a child? would an alternative be to pass in system and it can create a systemActorOf?

Reply:
or another level of indirection solves most problems ;-)

introducing a RememberEntitiesShardStoreProvider

```
trait RememberEntitiesShardStoreProvider {
  def actorResource: Option[Props]
  def createStore(factory: Option[ActorRef] => RememberEntitiesShardStore
```

then the actor would be created based on the `actorResource: Props`, and pass in that ref to the factory.

Reply:
Hmm, yeah, maybe something like that is the path to go. I'll give it a shot.